### PR TITLE
Training loop inference

### PR DESCRIPTION
**Pull request by gideonite**
Seems to be working. Check out this run: https://wandb.ai/ai2cm/fourcastnet-era5/runs/69lh5dh0?workspace=user-gideond

### CODE REVIEW

**Comment by gideonite** [comment]
To give a consistent wandb dashboard the x-axis is still training iteration, not epoch. But the epoch is logged along with the data so if we want to rescale the x-axis, that will be possible.

File: train.py [hunk 1]
```diff
@@ -198,13 +200,16 @@ def train(self):
 
     best_valid_loss = 1.e6
     for epoch in range(self.startEpoch, self.params.max_epochs):
+      logging.info("Epoch: %d" % epoch)
```
Comment (oliverwm1) [inline train.py:203]:
```suggestion
      logging.info("Epoch: %d" % epoch + 1)
```

1-based index probably makes more sense here, and is already the precedent set on line 241.

File: train.py [hunk 2]
```diff
@@ -198,13 +200,16 @@ def train(self):
...
       if dist.is_initialized():
         self.train_sampler.set_epoch(epoch)
 #        self.valid_sampler.set_epoch(epoch)
 
       start = time.time()
       tr_time, data_time, train_logs = self.train_one_epoch()
       valid_time, valid_logs = self.validate_one_epoch()
+      self.inference_one_epoch()
```
Comment (oliverwm1) [inline train.py:211]: I think the cleanest way to deal with the wandb step vs epoch issue would be to return the desired metrics/images as dicts from the `self.validate_one_epoch()` and `self.inference_one_epoch()` functions. Then these could be merged and passed to `wandb.log(...)` in a single call. This would require some refactoring of the `inference.py` script as well.

File: inference/inference.py [hunk 3]
```diff
@@ -321,9 +321,8 @@ def autoregressive_inference(params, ic, valid_data_full, model):
               **grad_mag_target_metrics
             }
           )
-              
 
-    if params.log_to_wandb:
+    if params.log_to_wandb and hasattr(params, 'epoch'):
```
Comment (oliverwm1) [inline inference/inference.py:325]: What if a user was just calling the `inference.py` script but still wanted to output these 5-day and 10-day metrics? Don't think we should require `epoch` to be in `params` here.

File: inference/inference.py [hunk 4]
```diff
@@ -337,7 +336,7 @@ def autoregressive_inference(params, ic, valid_data_full, model):
             for k in range(len(out_names)):
               name = f'{metric_names[i]}_{time_name}/ic{ic}/channel{k}-{out_names[k]}'
               try:
-                wandb.log({name: all_metrics[i, j, k]})
+                wandb.log({name: all_metrics[i, j, k], 'epoch': params.epoch}, step=params.iters)
```
Comment (oliverwm1) [inline inference/inference.py:339]: We could avoid having to specify step or add new 'epoch' metric if we log these to wandb outside of this function (see other comment).

**Review by oliverwm1 — CHANGES_REQUESTED** [review]
We're on the right track here, but still some changes required. I would like to avoid having to do any messing around with the x-axis in wandb.

Also make sure have an option to NOT do the wandb.log call for the metrics that are defined at every step of the inference (lines 301-323 in inference.py). We don't want to be logging these when doing online inference.

File: inference/inference.py (hunk 3 above)
Comment (gideonite) [reply inference/inference.py:328]: Right

File: train.py (hunk 2 above)
Comment (gideonite) [reply train.py]: sounds good

File: inference/inference.py (hunk 4 above)
Comment (gideonite) [reply inference/inference.py]: right

**Comment by gideonite** [comment]
Here's an example run with the updated code, https://wandb.ai/ai2cm/fourcastnet-era5/runs/vd8eaur2. @oliverwm1 

One of the new metrics:
image

Old metrics unchanged: 
image

File: inference/inference.py [hunk 5]
```diff
@@ -471,7 +476,8 @@ def autoregressive_inference(params, ic, valid_data_full, model):
     #run autoregressive inference for multiple initial conditions
     for i, ic in enumerate(ics):
       logging.info("Initial condition {} of {}".format(i+1, n_ics))
-      sr, sp, vl, a, au, vc, ac, acu, accland, accsea = autoregressive_inference(params, ic, valid_data_full, model)
+      sr, sp, vl, a, au, vc, ac, acu, accland, accsea, inference_logs = autoregressive_inference(params, ic, valid_data_full, model)
+      del inference_logs
```
Comment (oliverwm1) [inline inference/inference.py:480]:
```suggestion
      sr, sp, vl, a, au, vc, ac, acu, accland, accsea, _ = autoregressive_inference(params, ic, valid_data_full, model)
```
is typical for unused outputs.

File: train.py [hunk 6]
```diff
@@ -198,18 +200,20 @@ def train(self):
...
+      logging.info(f"Epoch: {epoch+1}")
       if dist.is_initialized():
         self.train_sampler.set_epoch(epoch)
 #        self.valid_sampler.set_epoch(epoch)
 
       start = time.time()
       tr_time, data_time, train_logs = self.train_one_epoch()
       valid_time, valid_logs = self.validate_one_epoch()
+      inference_logs = self.inference_one_epoch()
+      del tr_time, data_time, valid_time
```
Comment (oliverwm1) [inline train.py:212]: is this necessary, or just that these variables are unused?

**Review by oliverwm1 — COMMENTED** [review]
Thanks for changes, wandb run looks great! Only last request is that we not log the RMSE/ACC at each inference step to stdout if inference is happening during training, since it makes the training logs quite long/harder to read (IMO). You could use the same new param you defined.

File: inference/inference.py (hunk 3 above)
Comment (oliverwm1) [reply inference/inference.py:328]: Is the `hasattr(params, 'epoch')` still here intentionally? Don't think we need it?

File: inference/inference.py (hunk 5 above)
Comment (gideonite) [reply inference/inference.py:480]: Yes, I believe you are referring to PEP 640. I picked this up as an alternative way of handling unused variables [here](https://github.com/google/jax-cfd/blob/07e963417c27ce26ba272b31e555d6c21fdc3d5d/jax_cfd/ml/interpolations.py#L61). I like it because it documents the outputs of the function without forcing the user to find the return statement of an untyped function. The `del` is nice to document that the variable is unused (and save on memory).

File: train.py (hunk 6 above)
Comment (gideonite) [reply train.py:212]: These variables are unused. See comment above.

**Comment by gideonite** [comment]
Right, didn't consider that. Should be good to go!

File: inference/inference.py (hunk 5 above)
Comment (oliverwm1) [reply inference/inference.py:484]: I guess I just don't think that instances of the function call should be responsible for documenting its outputs. Especially if the variable is deleted right away, it's hard to know what its intended use is (I wouldn't put much weight in a placeholder name).

Anyway, doesn't need to hold up this PR—I just find the pattern a little confusing.

In the example you link, these deleted variables should just not be arguments to the init function...

**Review by oliverwm1 — APPROVED** [review]
Thanks for fixing up the logging!
